File: ainbox/global_init.py
"""Global agent integration bootstrap."""

from __future__ import annotations

from dataclasses import dataclass
import os
from pathlib import Path
import shutil
import subprocess


@dataclass(frozen=True)
class AgentIntegration:
    name: str
    executable: str
    marketplace_source: str
    marketplace_name: str
    plugins: tuple[str, ...]


SUPPORTED_INTEGRATIONS = (
    AgentIntegration(
        name="Claude Code",
        executable="claude",
        marketplace_source="JerrettDavis/AInbox",
        marketplace_name="ainbox-marketplace",
        plugins=(
            "ainbox@ainbox-marketplace",
            "agent-poll@ainbox-marketplace",
            "elections@ainbox-marketplace",
        ),
    ),
    AgentIntegration(
        name="GitHub Copilot CLI",
        executable="copilot",
        marketplace_source="JerrettDavis/AInbox",
        marketplace_name="ainbox-marketplace",
        plugins=(
            "ainbox@ainbox-marketplace",
            "agent-poll@ainbox-marketplace",
            "elections@ainbox-marketplace",
        ),
    ),
)

MAILBOX_MEMORY_TEXT = """# MAILBOX.md

Use AInbox as the durable coordination layer.

- Preserve or set `MAILBOX_AGENT_ID` explicitly.
- Use `.mailbox/draft/` as living memory for active threads.
- Sync before reading and after sending important updates.
- Use `mailbox send`, `list`, `read`, and `sync` for durable coordination.
- Reuse `--correlation-id` for related conversations.
- Use polls or elections when the group needs a durable decision or role assignment.
"""


def _run_command(command: list[str]) -> subprocess.CompletedProcess[str]:
    return subprocess.run(
        command,
        text=True,
        capture_output=True,
        check=False,
    )


def _resolved_executable(agent: AgentIntegration) -> str:
    override_name = f"MAILBOX_{agent.executable.upper()}_BIN"
    return os.environ.get(override_name, agent.executable)


def _command_exists(command: str) -> bool:
    path = Path(command)
    if path.is_absolute() or path.parent != Path():
        return path.is_file()
    return shutil.which(command) is not None


def _ensure_include_file(path: Path, include_line: str) -> bool:
    path.parent.mkdir(parents=True, exist_ok=True)
    if not path.exists():
        path.write_text(f"{include_line}\n", encoding="utf-8")
        return True

    content = path.read_text(encoding="utf-8")
    if include_line in content:
        return False

    path.write_text(f"{include_line}\n\n{content}", encoding="utf-8")
    return True


def _ensure_mailbox_file(path: Path) -> bool:
    path.parent.mkdir(parents=True, exist_ok=True)
    if path.exists():
        return False
    path.write_text(MAILBOX_MEMORY_TEXT.rstrip() + "\n", encoding="utf-8")
    return True


def _ensure_memory_install(root: Path, mailbox_rel: Path, include_target: Path, include_line: str, label: str) -> str:
    mailbox_created = _ensure_mailbox_file(root / mailbox_rel)
    include_updated = _ensure_include_file(root / include_target, include_line)
    actions = []
    if mailbox_created:
        actions.append(f"created {mailbox_rel}")
    if include_updated:
        actions.append(f"updated {include_target}")
    if not actions:
        actions.append("already configured")
    return f"{label}: " + ", ".join(actions)


def ensure_project_memory_files(root: Path | None = None) -> list[str]:
    project_root = (root or Path.cwd()).resolve()
    return [
        _ensure_memory_install(
            project_root,
            Path(".claude/MAILBOX.md"),
            Path("CLAUDE.md"),
            "@.claude/MAILBOX.md",
            "Claude Code project memory",
        ),
        _ensure_memory_install(
            project_root,
            Path(".agents/MAILBOX.md"),
            Path("AGENTS.md"),
            "@.agents/MAILBOX.md",
            "AGENTS project memory",
        ),
    ]


def ensure_user_memory_files() -> list[str]:
    home_root = Path.home()
    return [
        _ensure_memory_install(
            home_root,
            Path(".claude/MAILBOX.md"),
            Path(".claude/CLAUDE.md"),
            "@MAILBOX.md",
            "Claude Code user memory",
        ),
        _ensure_memory_install(
            home_root,
            Path(".agents/MAILBOX.md"),
            Path(".agents/AGENTS.md"),
            "@MAILBOX.md",
            "AGENTS user memory",
        ),
    ]


def _ensure_marketplace(agent: AgentIntegration) -> str:
    executable = _resolved_executable(agent)
    update = _run_command(
        [executable, "plugin", "marketplace", "update", agent.marketplace_name]
    )
    if update.returncode == 0:
        return "updated"

    add = _run_command(
        [executable, "plugin", "marketplace", "add", agent.marketplace_source]
    )
    if add.returncode == 0:
        return "added"

    raise RuntimeError(
        f"{agent.name}: failed to update or add marketplace {agent.marketplace_name}\n"
        f"update stderr: {update.stderr.strip()}\n"
        f"add stderr: {add.stderr.strip()}"
    )


def _ensure_plugin(agent: AgentIntegration, plugin: str) -> str:
    executable = _resolved_executable(agent)
    update = _run_command([executable, "plugin", "update", plugin])
    if update.returncode == 0:
        return "updated"

    install = _run_command([executable, "plugin", "install", plugin])
    if install.returncode == 0:
        return "installed"

    raise RuntimeError(
        f"{agent.name}: failed to update or install plugin {plugin}\n"
        f"update stderr: {update.stderr.strip()}\n"
        f"install stderr: {install.stderr.strip()}"
    )


def ensure_global_integrations() -> list[str]:
    """Install or update supported agent integrations."""
    summaries: list[str] = ensure_user_memory_files()

    for agent in SUPPORTED_INTEGRATIONS:
        executable = _resolved_executable(agent)
        if not _command_exists(executable):
            summaries.append(f"{agent.name}: skipped ({agent.executable} not found)")
            continue

        marketplace_status = _ensure_marketplace(agent)
        plugin_summaries = []
        for plugin in agent.plugins:
            plugin_summaries.append(f"{plugin} { _ensure_plugin(agent, plugin) }")

        summaries.append(
            f"{agent.name}: marketplace {marketplace_status}; "
            + ", ".join(plugin_summaries)
        )

    return summaries

File: src/global_init.rs
use crate::util::CliResult;
use std::env;
use std::fs;
use std::path::{Path, PathBuf};
use std::process::Command;

struct AgentIntegration {
    name: &'static str,
    executable: &'static str,
    marketplace_source: &'static str,
    marketplace_name: &'static str,
    plugins: &'static [&'static str],
}

const SUPPORTED_INTEGRATIONS: &[AgentIntegration] = &[
    AgentIntegration {
        name: "Claude Code",
        executable: "claude",
        marketplace_source: "JerrettDavis/AInbox",
        marketplace_name: "ainbox-marketplace",
        plugins: &[
            "ainbox@ainbox-marketplace",
            "agent-poll@ainbox-marketplace",
            "elections@ainbox-marketplace",
        ],
    },
    AgentIntegration {
        name: "GitHub Copilot CLI",
        executable: "copilot",
        marketplace_source: "JerrettDavis/AInbox",
        marketplace_name: "ainbox-marketplace",
        plugins: &[
            "ainbox@ainbox-marketplace",
            "agent-poll@ainbox-marketplace",
            "elections@ainbox-marketplace",
        ],
    },
];

const MAILBOX_MEMORY_TEXT: &str = r#"# MAILBOX.md

Use AInbox as the durable coordination layer.

- Preserve or set `MAILBOX_AGENT_ID` explicitly.
- Use `.mailbox/draft/` as living memory for active threads.
- Sync before reading and after sending important updates.
- Use `mailbox send`, `list`, `read`, and `sync` for durable coordination.
- Reuse `--correlation-id` for related conversations.
- Use polls or elections when the group needs a durable decision or role assignment.
"#;

pub fn ensure_project_memory_files(root: &Path) -> CliResult<Vec<String>> {
    Ok(vec![
        ensure_memory_install(
            root,
            Path::new(".claude/MAILBOX.md"),
            Path::new("CLAUDE.md"),
            "@.claude/MAILBOX.md",
            "Claude Code project memory",
        )?,
        ensure_memory_install(
            root,
            Path::new(".agents/MAILBOX.md"),
            Path::new("AGENTS.md"),
            "@.agents/MAILBOX.md",
            "AGENTS project memory",
        )?,
    ])
}

pub fn ensure_global_integrations() -> CliResult<Vec<String>> {
    let mut summaries = ensure_user_memory_files()?;

    for agent in SUPPORTED_INTEGRATIONS {
        let executable = resolved_executable(agent);
        if !command_exists(&executable) {
            summaries.push(format!(
                "{}: skipped ({} not found)",
                agent.name, agent.executable
            ));
            continue;
        }

        let marketplace_status = ensure_marketplace(agent, &executable)?;
        let mut plugin_summaries = Vec::new();
        for plugin in agent.plugins {
            plugin_summaries.push(format!(
                "{plugin} {}",
                ensure_plugin(agent, &executable, plugin)?
            ));
        }

        summaries.push(format!(
            "{}: marketplace {}; {}",
            agent.name,
            marketplace_status,
            plugin_summaries.join(", ")
        ));
    }

    Ok(summaries)
}

fn ensure_user_memory_files() -> CliResult<Vec<String>> {
    let home_root = user_home_dir()?;
    Ok(vec![
        ensure_memory_install(
            &home_root,
            Path::new(".claude/MAILBOX.md"),
            Path::new(".claude/CLAUDE.md"),
            "@MAILBOX.md",
            "Claude Code user memory",
        )?,
        ensure_memory_install(
            &home_root,
            Path::new(".agents/MAILBOX.md"),
            Path::new(".agents/AGENTS.md"),
            "@MAILBOX.md",
            "AGENTS user memory",
        )?,
    ])
}

fn ensure_memory_install(
    root: &Path,
    mailbox_rel: &Path,
    include_target: &Path,
    include_line: &str,
    label: &str,
) -> CliResult<String> {
    let mailbox_created = ensure_mailbox_file(&root.join(mailbox_rel))?;
    let include_updated = ensure_include_file(&root.join(include_target), include_line)?;
    let mut actions = Vec::new();
    if mailbox_created {
        actions.push(format!("created {}", mailbox_rel.display()));
    }
    if include_updated {
        actions.push(format!("updated {}", include_target.display()));
    }
    if actions.is_empty() {
        actions.push("already configured".to_string());
    }
    Ok(format!("{label}: {}", actions.join(", ")))
}

fn ensure_mailbox_file(path: &Path) -> CliResult<bool> {
    if path.exists() {
        return Ok(false);
    }
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent).map_err(|err| err.to_string())?;
    }
    fs::write(path, format!("{MAILBOX_MEMORY_TEXT}\n")).map_err(|err| err.to_string())?;
    Ok(true)
}

fn ensure_include_file(path: &Path, include_line: &str) -> CliResult<bool> {
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent).map_err(|err| err.to_string())?;
    }

    if !path.exists() {
        fs::write(path, format!("{include_line}\n")).map_err(|err| err.to_string())?;
        return Ok(true);
    }

    let content = fs::read_to_string(path).map_err(|err| err.to_string())?;
    if content.contains(include_line) {
        return Ok(false);
    }

    fs::write(path, format!("{include_line}\n\n{content}")).map_err(|err| err.to_string())?;
    Ok(true)
}

fn user_home_dir() -> CliResult<PathBuf> {
    env::var_os("USERPROFILE")
        .or_else(|| env::var_os("HOME"))
        .map(PathBuf::from)
        .filter(|path| !path.as_os_str().is_empty())
        .ok_or_else(|| "Could not determine user home directory".to_string())
}

fn resolved_executable(agent: &AgentIntegration) -> String {
    env::var(format!(
        "MAILBOX_{}_BIN",
        agent.executable.to_ascii_uppercase()
    ))
    .unwrap_or_else(|_| agent.executable.to_string())
}

fn ensure_marketplace(agent: &AgentIntegration, executable: &str) -> CliResult<&'static str> {
    let update = run_command(
        executable,
        &["plugin", "marketplace", "update", agent.marketplace_name],
    )?;
    if update.status.success() {
        return Ok("updated");
    }

    let add = run_command(
        executable,
        &["plugin", "marketplace", "add", agent.marketplace_source],
    )?;
    if add.status.success() {
        return Ok("added");
    }

    Err(format!(
        "{}: failed to update or add marketplace {}\nupdate stderr: {}\nadd stderr: {}",
        agent.name,
        agent.marketplace_name,
        String::from_utf8_lossy(&update.stderr).trim(),
        String::from_utf8_lossy(&add.stderr).trim()
    ))
}

fn ensure_plugin(
    agent: &AgentIntegration,
    executable: &str,
    plugin: &str,
) -> CliResult<&'static str> {
    let update = run_command(executable, &["plugin", "update", plugin])?;
    if update.status.success() {
        return Ok("updated");
    }

    let install = run_command(executable, &["plugin", "install", plugin])?;
    if install.status.success() {
        return Ok("installed");
    }

    Err(format!(
        "{}: failed to update or install plugin {}\nupdate stderr: {}\ninstall stderr: {}",
        agent.name,
        plugin,
        String::from_utf8_lossy(&update.stderr).trim(),
        String::from_utf8_lossy(&install.stderr).trim()
    ))
}

fn run_command(program: &str, args: &[&str]) -> CliResult<std::process::Output> {
    Command::new(program)
        .args(args)
        .output()
        .map_err(|err| err.to_string())
}

fn command_exists(program: &str) -> bool {
    let path = Path::new(program);
    if path.is_absolute()
        || path
            .parent()
            .is_some_and(|parent| !parent.as_os_str().is_empty())
    {
        return path.is_file();
    }

    env::var_os("PATH")
        .map(|paths| env::split_paths(&paths).any(|dir| executable_candidates(program, &dir)))
        .unwrap_or(false)
}

fn executable_candidates(program: &str, dir: &Path) -> bool {
    let base = dir.join(program);
    if base.is_file() {
        return true;
    }

    if cfg!(windows) {
        for ext in [".exe", ".cmd", ".bat"] {
            if dir.join(format!("{program}{ext}")).is_file() {
                return true;
            }
        }
    }

    false
}
